docs/_docs/2_2_writing_data.md

```diff
@@ -176,15 +176,49 @@ In some cases, you may want to migrate your existing table into Hudi beforehand.
 
 ## Datasource Writer
 
-The `hudi-spark` module offers the DataSource API to write (and also read) any data frame into a Hudi table.
-Following is how we can upsert a dataframe, while specifying the field names that need to be used
-for `recordKey => _row_key`, `partitionPath => partition` and `precombineKey => timestamp`
+The `hudi-spark` module offers the DataSource API to write (and read) a Spark DataFrame into a Hudi table. There are a number of options available:
 
+**`HoodieWriteConfig`**:
+
+**TABLE_NAME** (Required)<br>
+
+
+**`DataSourceWriteOptions`**:
+
+**RECORDKEY_FIELD_OPT_KEY** (Required): Primary key field(s). Nested fields can be specified using the dot notation eg: `a.b.c`. When using multiple columns as primary key use comma seperated notaion, eg: `"col1,col2,col3,etc"`. Single or multiple columns as primary key specified by `KEYGENERATOR_CLASS_OPT_KEY` property.<br>
+Default value: `"uuid"`<br>
+
+**PARTITIONPATH_FIELD_OPT_KEY** (Required): Columns to be used for partitioning the table. To prevent partitioning, provide empty string as value eg: `""`. Specify paritioning/no partitioning using `HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY`<br>
+Default value: `"partitionpath"`<br>
+
+**PRECOMBINE_FIELD_OPT_KEY** (Required): When two records have the same key value, the record with the largest value from the field specified here will be choosen.<br>
+Default value: `"ts"`<br>
+
+**OPERATION_OPT_KEY**: The [write operations](#write-operations) to use. Note: this cannot change across writes.<br>
```

Comment:
hi, why this cannot change across writes? would you please clarify? IIRU, It would be changed across writes.

Reply:
Good catch, that was meant to go in the description of the next parameter, TABLE_TYPE_OPT_KEY.

The text can be made more clear by using the following language instead
Instead of "Note: this cannot change across writes."
Use "Note: After the initial creation of a table, this value must stay consistent when writing to the table using the Spark SaveMode.Append mode."

Also, I believe that the key and partition columns, cannot be changed after the table is first created as well, so this note snippet can be added to those fields as well. Is this correct?

Reply:
Btw, there seems to be a mismatch in the data deletion documentation as well. To delete records it seems the only thing necessary is to set OPERATION_OPT_KEY to DELETE_OPERATION_OPT_VAL and HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY to `classOf[GlobalDeleteKeyGenerator].getCanonicalName`, however https://hudi.apache.org/docs/writing_data.html#deletes specifies a hard delete procedure that is different from this one.

Reply:
in fact, there are some ways to delete records in hudi, 1. we would use "org.apache.hudi.EmptyHoodieRecordPayload" to delete a batch of records, 2. we could directly use `delete` api, 3. we could we `_hoodie_is_delete` flag to delete record. Also the `HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY ` has nothing to do with the writing path IIRC, it is only related to hive sync.

Reply:
> Good catch, that was meant to go in the description of the next parameter, TABLE_TYPE_OPT_KEY.
> 
> The text can be made more clear by using the following language instead
> Instead of "Note: this cannot change across writes."
> Use "Note: After the initial creation of a table, this value must stay consistent when writing to (updating) the table using the Spark SaveMode.Append mode."
> 
> Also, I believe that the key and partition columns, cannot be changed after the table is first created as well, so this note snippet can be added to those fields as well. Is this correct?

Yes, `TABLE_TYPE_OPT_KEY` is not meant to be changed in next write, and `OPERATION_OPT_KEY` would be changed during different writes, and maybe I misunderstand what you meant.

Reply:
In regards to deletion: I believe method 2 that you mentioned with delete api, is the same thing as as setting  OPERATION_OPT_KEY to DELETE_OPERATION_OPT_VAL (DELETE_OPERATION_OPT_VAL has a string value of "delete"). In terms of method 3, I believe this flag method is only applicable when using the DeltaStreamer based on this blogpost https://hudi.apache.org/blog/delete-support-in-hudi/ , is this correct? And for method 1 of setting PAYLOAD_CLASS_OPT_KEY to "org.apache.hudi.EmptyHoodieRecordPayload", could you please clarify what the difference is between using method 1 and method 2? It seems the steps are the same as first a DataFrame must be prepared including the full records that need to be deleted, and then the write operation must be done using one of the 2 methods..

Reply:
the flag method is not only applicable when using DeltaStreamer, but also applicable when using HoodieWriteClient/Spark DataSource, in this way, the ingested batch record would consist of both new insert records and delete records, that means you would insert and delete records with one batch, but when setting to `EmptyHoodieRecordPayload`, the ingested batch record is considered to be all delete records.

Reply:
Understood. Will clarify the Deletes section with this.